design/workload-labeling.md

```diff
@@ -0,0 +1,236 @@
+            - containerPort: 80
+```
+
+In all, to achieve various goals, OAM users have to repeatedly specify information to manually establish connections
+between pods, workloads and traits.
+
+## Goals
+
+The goal is to recommend a simple way - Workload Labeling, to easily establish connections between Workloads, Components, AppConfigurations,
+and pods.
+
+## Proposal
+
+Step 1: Setting the label `definition.oam.dev/unique-pod` as the unique identifier of pods
+
+We found that if our final target workload resource is pod, we can always assume the pod has unique labels:
+
+- For pods created by K8s Deployment, the `pod-template-hash` label will be automatically added.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    pod-template-hash: 748f857667
+```
+
+- For pods created by K8s DaemonSet, StatefulSet, [OpenKruise/Cloneset](https://openkruise.io/en-us/docs/cloneset.html), `controller-revision-hash` label will be generated.
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    controller-revision-hash: b8d6c88f6
+```
+
+Declare `definition.oam.dev/unique-pod` in WorkloadDefinition as OAM specified labels, which will always exist in pods.
+Then trait automatically fills correlation fields by detecting the unique pod label, for example, generating service, split traffic, etc.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: WorkloadDefinition
+metadata:
+  name: xxx
+  labels:
+    definition.oam.dev/unique-pod: pod-template-hash # optional but recommended, values in [pod-template-hash, controller-revision-hash]
+spec:
+  ...
+```
+
+Step 2: Utilizing generated workload labels to establish clear relationship between workloads and its component
+
+OAM Kubernetes Runtime now has the ability of [automatically generating these labels for every workload](https://github.com/crossplane/oam-kubernetes-runtime/pull/189) as below.
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  annotations:
+    deployment.kubernetes.io/revision: "1"
+  labels:
+    app: nginx
+    app.oam.dev/component: label-component
+    app.oam.dev/name: label-appconfig
+    app.oam.dev/revision: label-component-v1
+    definition.oam.dev/unique-pod: pod-template-hash
+```
+
+Label `app.oam.dev/component` marks the Component, `app.oam.dev/name` marks the ApplicationConfiguration, and `app.oam.dev/revision`
+marks the revision of the component.
+
+Currently, a trait can find the attached workload by [`spec.workloadRef`](./one-pager-trait-scope-workload-interaction-mechanism.md) field of the Trait. With those labels above, a
+trait can quickly find the corresponding component and its revision.
+
+```yaml
+spec:
+  workloadRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: label-nginx
+```
+
+Moreover, with recommended label `definition.oam.dev/unique-pod` in Component manifest set, we can see it is
+generated as one label of the Deployment workload, which can help identify the pods.
+
+## Best practice
+
+- For OAM users
```

Comment:
I don't like using `OAM users` here, it can't be understood clearly.

We general distinguish users by

* Platform builder
* Application Operator
* Application Developer